### PR TITLE
Better error handling using `do`

### LINKED ISSUE #4310: Better handling of errors
The problem of error reporting is two-fold: First, we need to detect an error, then have a way to react to it. Currently, if there is a ShellError (not talking about ParseErrors which IMO should always be reported), it will be reported at the pipeline boundary and the script terminates.

To capture an error we could do something like 
```
> do --capture-error { blah }
```
which  would return `Value::Error` instead of direct `ShellError`, so in theory, it should get passed as a value and not raise the error.

The second part is to have a command that detects `Value::Error` so we could have something like this:
```
> do --capture-error { blah } | if error? {
    oops
} else {
    we're fine
}
```

I think we can build more on top of that, for example have an explicit error type so we build error handlers:
```
> def handle-error [err: error] {
   ...
}
```

The errors could also be structured data so you can check the error type, error message, etc.

### PR DESCRIPTION
**Pull request by merelymyself**
# Description

Attempts to fix #4310. Adds a flag to `do` to collect errors. Adds a type-check command as well for easy type checking. 

# Tests

Make sure you've done the following:

- [ ] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [ ] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [ ] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [ ] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [ ] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [ ] `cargo test --workspace --features=extra` to check that all the tests pass

### CODE REVIEW

**Comment by fdncred** [comment]
I've idea!

File: crates/nu-protocol/src/syntax_shape.rs [hunk 1]
```diff
@@ -112,6 +115,7 @@ impl SyntaxShape {
             SyntaxShape::Filesize => Type::Filesize,
             SyntaxShape::FullCellPath => Type::Any,
             SyntaxShape::GlobPattern => Type::String,
+            SyntaxShape::Error => Type::Error,
```
Comment (fdncred) [inline crates/nu-protocol/src/syntax_shape.rs:118]: if we're adding `SyntaxShape::Error`, I'm kind of wondering if it needs to be in other places as well. Here are some places I question.

1. parser.rs parse_shapename?
2. parser.rs parse_type?
3. ty.rs maybe a new Type?
4. ty.rs to_shape?
5. ty.rs fmt?
6. others?

Comment (merelymyself) [reply crates/nu-protocol/src/syntax_shape.rs:118]: The ones in `ty.rs` were already completed, I was pleasantly surprised to discover. I think the ones you mentioned in `parser.rs` have also been dealt with, but I'll keep a lookout for other areas.

Comment (fdncred) [reply crates/nu-protocol/src/syntax_shape.rs:118]: ok, i see them now except for in parse_shape_name.

Comment (fdncred) [reply crates/nu-protocol/src/syntax_shape.rs:118]: oh sorry. I think I'm mistaken.

**Comment by fdncred** [comment]
let's try this. thanks.

**Pull request merged**